### PR TITLE
pgplot: Add intel compiler support and add more driver

### PR DESCRIPTION
**Pull request by jasonfenglu**
1. It can be compiled with intel compiler. 
2. Enable drivers for X Window, PS, PNG and Latex. 

Tested with intel compiler 2021.2.0, gcc@4.8.5 and gcc@11.2.0

### CODE REVIEW

**Comment by jasonfenglu** [comment]
I am not sure how to do with the variants. 
It is common to use PGPLOT to plot PNG file or draw on X Window,
but then the package **lorene** and should change the default _depend_on_ behavior.

**Comment by alalazo** [comment]
Waiting on a review from @eschnett

**Comment by eschnett** [comment]
@jasonfenglu How would you change the `pgplot` dependency in `lorene`? `lorene` doesn't really care which features are enabled by default.

My intuition is that enabling `png` and `ps` by default is fine, but enabling `X` might pull in too many dependencies to be a good default. (Am I wrong here?) I'm assuming here that people will use most packages non-interactively by default.

**Comment by jasonfenglu** [comment]
I agree that this package should be non-interaction environment friendly. 
What worries me is if `png` is set as default, then the dependency tree will grow when you build `lorene` (libpng and zlib). 
If @eschnett you think it is ok then I will just set `png` and `ps` as default and `X` as an option. 
I am not sure why I add `libx11` dependency when building with `png` in the first place and it will be removed.

File: var/spack/repos/builtin/packages/pgplot/package.py
Comment (eschnett) [inline var/spack/repos/builtin/packages/pgplot/package.py]: Is `type=('build', 'run')` necessary here? Wouldn't the default work fine?

**Review by eschnett — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/pgplot/package.py
Comment (jasonfenglu) [reply var/spack/repos/builtin/packages/pgplot/package.py]: It is required in the first place or `libpng` is not found when executing pgdemos, but I can't reproduce that problem now. Maybe it solved by some modification to the makefile automatically (like the zlib in the next commit) . I'll remove it and see how that goes.

**Comment by jasonfenglu** [comment]
`zlib` has to be specified in the linking flag.

File: var/spack/repos/builtin/packages/pgplot/package.py
Comment (eschnett) [inline var/spack/repos/builtin/packages/pgplot/package.py]: Does `pgplot` use zlib directly, or is this just a transitive dependency because `libpng` uses `zlib`?

Comment (jasonfenglu) [reply var/spack/repos/builtin/packages/pgplot/package.py]: `pgplot` does not use zlib. `libpng` dependes on `zlib` so I have to explicitly tell makefile to include it. The compiler wrapper by spack should have taken care of it but it is not.

Comment (eschnett) [reply var/spack/repos/builtin/packages/pgplot/package.py]: It would be better to write
```Python
sub['@CCOMPL@'] += ' {0}'.format(self.spec['libpng'].headers.cpp_flags)
sub['@FFLAGD@'] += ' {0}'.format(self.spec['libpng'].libs.ld_flags)
```
if that works. I hope that `libpng`'s dependencies are included here. If so, the `depends_on('zlib')` can also be removed.

Comment (jasonfenglu) [reply var/spack/repos/builtin/packages/pgplot/package.py]: I think it is not working if I change to it to this:

```diff
         if '+png' in spec:
             enable_driver('! PNDRIV 1 /PNG')
-            sub['@CCOMPL@'] += ' -I{0}'.format(self.spec['libpng'].prefix.include)
-            sub['@FFLAGD@'] += ' -L{0} -lpng'.format(self.spec['libpng'].prefix.lib)
-            sub['@FFLAGD@'] += ' -L{0} -lz'.format(self.spec['zlib'].prefix.lib)
+            sub['@CCOMPL@'] += ' {0}'.format(self.spec['libpng'].headers.cpp_flags)
+            sub['@FFLAGD@'] += ' {0}'.format(self.spec['libpng'].libs.ld_flags)

             filter_file('pndriv.o : ./png.h ./pngconf.h ./zlib.h ./zconf.h',
-                        'pndriv.o : {0}/png.h {0}/pngconf.h {1}/zlib.h {1}/zconf.h'.
-                        format(self.spec['libpng'].prefix.include,
-                               self.spec['zlib'].prefix.include), 'makemake')
+                        'pndriv.o :', 'makemake')
+
             self.rpath.append(self.spec['libpng'].prefix)
```

And the replaced `g77_gcc.conf` file contains no information of zlib like this:
```   
   FFLAGD="-nofor-main -L/work/packages/spack_20200621/opt/spack/linux-centos7-westmere/intel-2021.2.0/libpng-1.6.37-sr7zyzuno7kv2wnh5kqjycgs62o7puze/lib -lpng"

# Mandatory.
# The C compiler to use.

   CCOMPL="/work/packages/spack_20200621/opt/spack/linux-centos7-nehalem/gcc-4.8.5/intel-oneapi-compilers-2021.2.0-muplh6xossdrscjp7uzrjssunymlvinr/compiler/2021.2.0/linux/bin/intel64/icc -I/work/packages/spack_20200621/opt/spack/linux-centos7-westmere/intel-2021.2.0/libpng-1.6.37-sr7zyzuno7kv2wnh5kqjycgs62o7puze/include"
```

Comment (eschnett) [reply var/spack/repos/builtin/packages/pgplot/package.py]: Thanks for trying.

**Review by eschnett — DISMISSED** [review]

**Comment by jasonfenglu** [comment]
Hi @eschnett , I found a way to include `zlib` related flags from `libpng` spec. I guess it is what you are looking for.  `depend_on(zlib, when='+png')` is no longer needed.

**Comment by eschnett** [comment]
@jasonfenglu Thanks!

**Review by eschnett — APPROVED** [review]

**Review by alalazo — APPROVED** [review]
Maintainer approved so LGTM

**Pull request merged**
